### PR TITLE
HDDS-2325. BenchMarkDatanodeDispatcher genesis test is failing with NPE

### PR DESCRIPTION
**Pull request by elek**
## What changes were proposed in this pull request?

TLDR; Fix BenchmarkDatanodeDispatcher.writeChunk test.

Genesis is a microbenchmark tool for Ozone based on JMH (https://openjdk.java.net/projects/code-tools/jmh/).

Due to the recent Datanode changes the BenchMarkDatanodeDispatcher is failing with NPE:

```
java.lang.NullPointerException
	at org.apache.hadoop.ozone.container.common.interfaces.Handler.<init>(Handler.java:69)
	at org.apache.hadoop.ozone.container.keyvalue.KeyValueHandler.<init>(KeyValueHandler.java:114)
	at org.apache.hadoop.ozone.container.common.interfaces.Handler.getHandlerForContainerType(Handler.java:78)
	at org.apache.hadoop.ozone.genesis.BenchMarkDatanodeDispatcher.initialize(BenchMarkDatanodeDispatcher.java:115)
	at org.apache.hadoop.ozone.genesis.generated.BenchMarkDatanodeDispatcher_createContainer_jmhTest._jmh_tryInit_f_benchmarkdatanodedispatcher0_G(BenchMarkDatanodeDispatcher_createContainer_jmhTest.java:438)
	at org.apache.hadoop.ozone.genesis.generated.BenchMarkDatanodeDispatcher_createContainer_jmhTest.createContainer_Throughput(BenchMarkDatanodeDispatcher_createContainer_jmhTest.java:71)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at org.openjdk.jmh.runner.BenchmarkHandler$BenchmarkTask.call(BenchmarkHandler.java:453)
	at org.openjdk.jmh.runner.BenchmarkHandler$BenchmarkTask.call(BenchmarkHandler.java:437)
	at java.util.concurrent.FutureTask.run(FutureTask.java:266)
	at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511)
	at java.util.concurrent.FutureTask.run(FutureTask.java:266)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
	at java.lang.Thread.run(Thread.java:748)
 ```

And this is the just the biggest problem there are a few other problems. I propose the following fixes:

**fix 1**: NPE is thrown because the 'context' object is required by KeyValueHandler/Handler classes.

In fact the context is not required, we need two functionalities/info from the context: the ability to send icr (IncrementalContainerReport) and the ID of the datanode.

Law of Demeter principle suggests to have only the minimum required information from other classes.

For example instead of having context but using only context.getParent().getDatanodeDetails().getUuidString() we can have only the UUID string which makes more easy to test (unit and benchmark) the Handler/KeyValueHandler.

This is the biggest (but still small change) in this patch: I started to use the datanodeId and an icrSender instead of having the full context.

**fix 2,3**: There were a few other problems. The scmId was missing if the writeChunk was called from Benchmark and and the Checksum was also missing.

**fix 4**: I also had a few other problems: very huge containers are used (default 5G) and as the benchmark starts with creating 100 containers it requires 500G space by default. I adjusted the container size to make it possible to run on local machine.

## What is the link to the Apache JIRA

https://issues.apache.org/jira/browse/HDDS-2325

## How was this patch tested?

```
./ozone genesis -benchmark=BenchMarkDatanodeDispatcher.writeChunk
```

### CODE REVIEW

**Comment by elek** [comment]
notification test (notification is supposed to be sent to ozone-issues@hadoop.apache.org)

File: hadoop-hdds/container-service/src/main/java/org/apache/hadoop/ozone/container/ozoneimpl/OzoneContainer.java [hunk 1]
```diff
@@ -99,9 +105,21 @@ public OzoneContainer(DatanodeDetails datanodeDetails, OzoneConfiguration
     final ContainerMetrics metrics = ContainerMetrics.create(conf);
     this.handlers = Maps.newHashMap();
     for (ContainerType containerType : ContainerType.values()) {
+
+      Consumer<ContainerReplicaProto> icrSender = containerReplicaProto -> {
```
Comment (adoroszlai) [inline hadoop-hdds/container-service/src/main/java/org/apache/hadoop/ozone/container/ozoneimpl/OzoneContainer.java]: Wouldn't a single `icrSender` created outside the loop be enough?

File: hadoop-hdds/container-service/src/test/java/org/apache/hadoop/ozone/container/common/impl/TestHddsDispatcher.java [hunk 2]
```diff
@@ -98,8 +100,10 @@ public void testContainerCloseActionWhenFull() throws IOException {
       Map<ContainerType, Handler> handlers = Maps.newHashMap();
       for (ContainerType containerType : ContainerType.values()) {
         handlers.put(containerType,
-            Handler.getHandlerForContainerType(containerType, conf, context,
-                containerSet, volumeSet, metrics));
+            Handler.getHandlerForContainerType(containerType, conf,
+                context.getParent().getDatanodeDetails().getUuidString(),
+                containerSet, volumeSet, metrics, c -> {
+                }));
```
Comment (adoroszlai) [inline hadoop-hdds/container-service/src/test/java/org/apache/hadoop/ozone/container/common/impl/TestHddsDispatcher.java]: Nit: odd spacing for no-op consumer ([this one](https://github.com/apache/hadoop-ozone/pull/60/files#diff-d696db1fc042e05bbf8d305d5182f260R262) looks cleaner).  Same in a few other places.

File: hadoop-hdds/container-service/src/main/java/org/apache/hadoop/ozone/container/common/impl/HddsDispatcher.java [hunk 3]
```diff
@@ -304,6 +306,11 @@ private ContainerCommandResponseProto dispatchRequest(
         // Once container is marked unhealthy, all the subsequent write
         // transactions will fail with UNHEALTHY_CONTAINER exception.
 
+        if (container == null) {
+          throw new RuntimeException(
```
Comment (adoroszlai) [inline hadoop-hdds/container-service/src/main/java/org/apache/hadoop/ozone/container/common/impl/HddsDispatcher.java]: May be better to throw a more specific exception (`IllegalState` or `NullPointer`)?

File: hadoop-hdds/container-service/src/test/java/org/apache/hadoop/ozone/container/common/impl/TestHddsDispatcher.java (hunk 2 above)
Comment (elek) [reply hadoop-hdds/container-service/src/test/java/org/apache/hadoop/ozone/container/common/impl/TestHddsDispatcher.java]: Hard to avoid this as this is the default formatting rule of IntelliJ. Will be reformatted by the next auto-format anyway.

But I created a constant field, hopefully it makes it a little more readable:

```
  public static final Consumer<ContainerReplicaProto> NO_OP_ICR_SENDER = c -> {};

```

**Comment by elek** [comment]
Thanks the review @arp7 and @adoroszlai 

I re-tested it locally and committed it to the master branch.
